File: src/tractor/Player.java
package tractor;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Player extends SortedCardCollection {
	int id, level;

	public Player(int i) {
		super();
		id = i;
		level = 0;
	}

	public void levelUp(int level) {
		this.level += level;
	}

	public boolean hasComboToPlay(Combo firstCombo) {
		// TODO: refactor
		// has to be pair or tractor
		// the hard part is that tractors have multiple
		// easiest is to find biggest number of consecutive pairs in the suit
		// TODO make sure everything works with SHUAI
		int currentCount = 0;
		int maxCount = 0;
		int startOfSuit = 0;
		// find start of suit
		for (int i = 0; i < size(); i++) {
			if (get(i).getSuit() == firstCombo.getSuit()) {
				startOfSuit = i;
				break;
			}
		}
		int i = startOfSuit;
		int lastPair = -1;
		while (i < size() - 1 && get(i).getSuit() == firstCombo.getSuit()) {
			if (get(i).equals(get(i + 1))) {// is a pair
				if (lastPair == get(i).getPowerIndex()) {
					i += 2;
					continue;
				} else if (lastPair == -1 || get(i).getPowerIndex() != lastPair + 1) {
					// start of new pair list
					currentCount = 1;

				} else { // continuing list of consecutive
					currentCount++;
				}
				if (currentCount > maxCount) {
					maxCount = currentCount;
				}
				lastPair = get(i).getPowerIndex();
				i += 2;
			} else { // not a pair
				i++;
			}
		}
		if (maxCount * 2 >= firstCombo.size()) {
			return true;
		}
		return false;
	}

	public boolean playedRequired(Combo firstCombo, Combo attemptedCombo) {
		// TODO: refactor
		List<Integer> handConsecPairCounts = consecutivePairCounts(firstCombo);
		List<Integer> comboConsecPairCounts = attemptedCombo.consecutivePairCounts(firstCombo);
		int numberOfCardsNeeded = firstCombo.size();
		int i = 0;
		while (numberOfCardsNeeded != 0 && i < handConsecPairCounts.size()) {
			if (i >= comboConsecPairCounts.size()) {
				return false;
			}
			if (handConsecPairCounts.get(i) * 2 <= numberOfCardsNeeded) {
				if (comboConsecPairCounts.get(i) != handConsecPairCounts.get(i)) {
					return false;
				}
				numberOfCardsNeeded -= handConsecPairCounts.get(i);
			} else {
				if (comboConsecPairCounts.get(i) != numberOfCardsNeeded) {
					return false;
				} else {
					return true;
				}
			}
			i++;
		}
		return true;
	}

	public List<Integer> consecutivePairCounts(Combo firstCombo) {
		// TODO: refactor
		List<Integer> consecPairCounts = new ArrayList<Integer>();
		int currentCount = 0;
		int maxCount = 0;
		int startOfSuit = 0;
		// find start of suit
		for (int i = 0; i < size(); i++) {
			if (get(i).getSuit() == firstCombo.getSuit()) {
				startOfSuit = i;
				break;
			}
		}
		int i = startOfSuit;
		int lastPair = -1;
		while (i < size() - 1 && get(i).getSuit() == firstCombo.getSuit()) {
			if (get(i).equals(get(i + 1))) {// is a pair
				if (lastPair == get(i).getPowerIndex()) {
					i += 2;
					continue;
				} else if (lastPair == -1 || get(i).getPowerIndex() != lastPair + 1) {
					// start of new pair list
					currentCount = 1;

				} else { // continuing list of consecutive
					currentCount++;
				}
				if (currentCount > maxCount) {
					maxCount = currentCount;
				}
				lastPair = get(i).getPowerIndex();
				i += 2;
			} else { // not a pair
				i++;
				if (currentCount > 0) {
					consecPairCounts.add(currentCount);
				}
				currentCount = 0;
			}

		}
		if (currentCount > 0) {
			consecPairCounts.add(currentCount);
		}
		Collections.sort(consecPairCounts);
		Collections.reverse(consecPairCounts);
		return consecPairCounts;
	}

	public boolean canOverride(Player player, int newCallLevel) {
		// TODO: refactor
		int numCards = newCallLevel / 6 + 1;
		int suit = newCallLevel % 6;
		Card toCheck = null;
		if (suit == 5) {
			toCheck = new Card("J+");
			if (numCards == 1) {
				return false;
			}
		} else if (suit == 4) {
			toCheck = new Card("J-");
			System.out.println("blah!");
			if (numCards == 1) {
				return false;
			}
		} else {
			toCheck = new Card(Card.SUITS[suit] + Card.VALUES[Card.trumpValue]);
		}
		for (Card i : this) {
			if (i.equals(toCheck)) {
				numCards--;
			}
		}
		return numCards <= 0;
	}

	public boolean has(Card card, int numCards) {
		return getCount(card) >= numCards;
	}

}

File: src/python/Player.py
from SortedCardCollection import SortedCardCollection
class Player(SortedCardCollection):

  def __init__(self, i):
    super().__init__()
    self.id = i
    self.level = 0
    
  def levelUp(self, level):
    self.level += level
    
  def hasComboToPlay(self, firstCombo):
    currentCount = 0
    maxCount = 0
    startOfSuit = 0
    for i in range(self.size()):
      if self.get(i).getSuit() == firstCombo.getSuit():
        startOfSuit = i
        break
    i = startOfSuit
    lastPair = -1
    while i < self.size() - 1 and self.get(i).getSuit() == firstCombo.getSuit():
      if self.get(i) == self.get(i + 1):
        if lastPair == self.get(i).getPowerIndex():
          i += 2
          continue
        elif lastPair == -1 or self.get(i).getPowerIndex() != lastPair + 1:
          currentCount = 1
        else:
          currentCount += 1
        if currentCount > maxCount:
          maxCount = currentCount
        lastPair = self.get(i).getPowerIndex()
        i += 2
      else: 
        i += 1
    if maxCount * 2 >= firstCombo.size():
      return True
    return False
  
  def playedRequired(self, firstCombo, attemptedCombo):
    handConsecPairCounts = self.consecutivePairCounts(firstCombo)
    comboConsecPairCounts = attemptedCombo.consecutivePairCounts(firstCombo)
    numberOfCardsNeeded = firstCombo.size()
    i = 0
    while numberOfCardsNeeded != 0 and i < len(handConsecPairCounts):
      if i >= len(comboConsecPairCounts):
        return False
      if handConsecPairCounts[i] * 2 <= numberOfCardsNeeded:
        if comboConsecPairCounts[i] != handConsecPairCounts[i]:
          return False
        numberOfCardsNeeded -= handConsecPairCounts[i]
      else:
        if comboConsecPairCounts[i] != numberOfCardsNeeded:
          return False
        else:
          return True
      i+=1
    return True
  
  def consecutivePairCounts(self, firstCombo):
    consecPairCounts = []
    currentCount = 0
    maxCount = 0
    startOfSuit = 0
    for i in range(self.size()):
      if self.get(i).getSuit() == firstCombo.getSuit():
        startOfSuit = i
        break
    i = startOfSuit
    lastPair = -1
    while i < self.size() - 1 and self.get(i).getSuit() == firstCombo.getSuit():
      if self.get(i) == self.get(i + 1):
        if lastPair == self.get(i).getPowerIndex():
          i += 2
          continue
        elif lastPair == -1 or self.get(i).getPowerIndex() != lastPair + 1:
          currentCount = 1
        else:
          currentCount += 1
        if currentCount > maxCount:
          maxCount = currentCount
        lastPair = self.get(i).getPowerIndex()
        i += 2
      else: 
        i+=1
        if currentCount > 0:
          consecPairCounts.add(currentCount)
        currentCount = 0
    if currentCount > 0:
      consecPairCounts.append(currentCount)
    consecPairCounts.sort()
    consecPairCounts.reverse()
    return consecPairCounts
  
  def canOverride(self, player, newCallLevel):
    numCards = int(newCallLevel / 6) + 1
    suit = newCallLevel % 6
    toCheck = None
    if suit == 5:
      toCheck = Card("J+")
      if numCards == 1:
        return False
    elif suit == 4:
      toCheck = Card("J-")
      if numCards == 1:
        return False
    else:
      toCheck = Card(Card.SUITS[suit] + Card.VALUES[Card.trumpValue])
    for i in self:
      if i == toCheck:
        numCards-=1
    return numCards <= 0
  
  def has(self, card, numCards):
    return self.getCount(card) >= numCards
